BitFaster.Caching/Lru/TemplateConcurrentLru.cs

```diff
@@ -295,24 +295,106 @@ public void AddOrUpdate(K key, V value)
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="itemCount"/> is greater than capacity./</exception>
+        /// <remarks>
+        /// Note: Trim affects LRU order. Calling Trim resets the internal accessed status of items.
+        /// </remarks>
+        public void Trim(int itemCount)
+        {
+            int capacity = this.coldCapacity + this.warmCapacity + this.hotCapacity;
+
+            if (itemCount < 0 || itemCount > capacity)
+            { 
+                throw new ArgumentOutOfRangeException(nameof(itemCount), "itemCount must be greater than or equal to zero, and less than the capacity of the cache.");
+            }
+
+            // clamp itemCount to number of items actually in the cache
+            itemCount = Math.Min(itemCount, this.HotCount + this.WarmCount + this.ColdCount);
+
+            // first scan each queue for discardable items and remove them immediately. Note this can remove > itemCount items.
+            int itemsRemoved = TrimAllDiscardedItems();
 
-            // At this point, dictionary is empty but queues still hold references to all values.
-            // Cycle the queues to purge all refs. If any items were added during this process, 
-            // it is possible they might be removed as part of CycleCold. However, the dictionary
-            // and queues will remain in a consistent state.
-            for (int i = 0; i < keys.Count; i++)
+            TrimLiveItems(itemsRemoved, itemCount);
+        }
+
+        private int TrimAllDiscardedItems()
+        {
+            int itemsRemoved = 0;
+
+            void RemoveDiscardableItems(ConcurrentQueue<I> q, ref int queueCounter)
             {
-                CycleHotUnchecked();
-                CycleWarmUnchecked();
-                CycleColdUnchecked();
+                int localCount = queueCounter;
+
+                for (int i = 0; i < localCount; i++)
+                {
+                    if (q.TryDequeue(out var item))
+                    {
+                        if (this.itemPolicy.ShouldDiscard(item))
+                        {
+                            Interlocked.Decrement(ref queueCounter);
+                            this.Move(item, ItemDestination.Remove, ItemRemovedReason.Trimmed);
+                            itemsRemoved++;
+                        }
+                        else
+                        {
+                            q.Enqueue(item);
+                        }
+                    }
+                }
+            }
+
+            RemoveDiscardableItems(coldQueue, ref this.coldCount);
+            RemoveDiscardableItems(warmQueue, ref this.warmCount);
+            RemoveDiscardableItems(hotQueue, ref this.hotCount);
+
+            return itemsRemoved;
+        }
+
+        private void TrimLiveItems(int itemsRemoved, int itemCount)
+        {
+            while (itemsRemoved < itemCount)

```

Comment:
This could be an infinite loop - if another thread clears the cache and no further items are added.